### PR TITLE
add testcases for actions

### PR DESCRIPTION
**Pull request by Dinaamagdy**

### CODE REVIEW

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 1]
```diff
@@ -42,29 +43,29 @@ def test001_create_cloudspace_with_wrong_params(self):
         self.log('%s STARTED' % self._testID)
 
         self.log('Create an cloudspace with providing non existing parameter, should fail')
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: - an cloudspace --> a cloudspace
- Test case for creating "acloudspace" --> "a cloudspace"

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 2]
```diff
@@ -42,29 +43,29 @@ def test001_create_cloudspace_with_wrong_params(self):
...
         res = self.create_cs(openvcloud=self.openvcloud, accounts=self.accounts, vdcusers=self.vdcusers,
                              cloudspaces=self.cloudspaces, temp_actions=self.temp_actions)
-        self.assertEqual(res,"account doesn't exist" )
+        self.assertEqual(res, "account doesn't exist")
 
         self.log('Create an cloudspace with providing non existing user, should fail')
         self.cloudspaces[self.cs1] = {'account': self.acc1,
                                       'users': OrderedDict([('name', self.random_string()),
-                                                          ('accesstype', 'CXDRAU')])}
+                                                            ('accesstype', 'CXDRAU')])}
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: Adding users here is not applicable anymore.. please make sure it is covered in the action section

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 3]
```diff
@@ -42,29 +43,29 @@ def test001_create_cloudspace_with_wrong_params(self):
...
         self.assertIn('no vdcuser found', res)
 
         self.log('Create an cloudspace without providing account parameter, should fail')
         self.cloudspaces[self.cs1] = {}
         res = self.create_cs(openvcloud=self.openvcloud, vdcusers=self.vdcusers, accounts=self.accounts,
-                       cloudspaces=self.cloudspaces, temp_actions=self.temp_actions)
+                             cloudspaces=self.cloudspaces, temp_actions=self.temp_actions)
         self.assertEqual(res, 'account is required')
 
         self.log('%s ENDED' % self._testID)
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: --> Test case for testing multiple cloud spaces creation.
or 
--> Test case for creating more than one cloud space

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 4]
```diff
@@ -84,17 +85,26 @@ def test002_create_cloudspaces(self):
...
+                             cloudspaces=self.cloudspaces, temp_actions=self.temp_actions)
+        self.wait_for_service_action_status(self.cs1, res[self.cs1]['install'])
 
         self.log('Check that the cloudspaces have been created.')
         cloudspace1 = self.get_cloudspace(self.cs1)
         self.assertEqual(cloudspace1['status'], 'DEPLOYED')
         cloudspace2 = self.get_cloudspace(cs2)
         self.assertEqual(cloudspace2['status'], 'DEPLOYED')
 
+        self.log(' Uninstall an cloudspace [CS1], should succeed.')
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: --> a cloudspace

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 5]
```diff
@@ -84,17 +85,26 @@ def test002_create_cloudspaces(self):
...
+        temp_actions = {'vdc': {'actions': ['uninstall']}}
+        res = self.create_cs(openvcloud=self.openvcloud, vdcusers=self.vdcusers, accounts=self.accounts,
+                             cloudspaces=self.cloudspaces, temp_actions=temp_actions)
+        self.wait_for_service_action_status(self.cs1, res[self.cs1]['uninstall'])
+    
+        self.log("Check that cloudspace [CS1] uninstall sucessfuly , should succeed.")
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: --> successfully

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 6]
```diff
@@ -138,6 +148,259 @@ def test003_create_cloudspace_with_different_limitaions(self, type, factor):
...
+
+class CloudspaceActions(OVC_BaseTest):
+    def __init__(self, *args, **kwargs):
+        super(CloudspaceActions, self).__init__(*args, **kwargs)
+
+    @classmethod
+    def setUpClass(cls):
+        self = cls()
+        super(CloudspaceActions, self).setUp()
+        cls.acc1 = str(uuid.uuid4())[0:8]
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: please use --> self.random_string()

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 7]
```diff
@@ -138,6 +148,259 @@ def test003_create_cloudspace_with_different_limitaions(self, type, factor):
...
+        cls.cloudspaces = {cls.cs1: {'account': cls.acc1}}
+
+        cls.temp_actions = {'account': {'actions': ['install']},
+                            'vdcuser': {'actions': ['install']},
+                            'vdc': {'actions': ['install']}
+                            }
+        cls.cloudspaces[cls.cs1] = {'account': cls.acc1, 'maxMemoryCapacity': randint(10, 1000),
+                                    'maxCPUCapacity': randint(10, 1000), 'maxDiskCapacity':randint(10, 1000),                      
+                                    'maxNumPublicIP': randint(10, 1000), 'maxNetworkPeerTransfer':randint(10, 1000),
+                                    'users': OrderedDict([('name', cls.vdcuser),
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: users shouldn't be here

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 8]
```diff
@@ -138,6 +148,259 @@ def test003_create_cloudspace_with_different_limitaions(self, type, factor):
             self.assertEqual(cloudspace['resourceLimits']['CU_I'], CU_I)
             self.assertEqual(cloudspace['resourceLimits']['CU_M'], CU_M)
             self.assertEqual(cloudspace['resourceLimits']['CU_NP'], CU_NP)
-            self.assertIn('gig_qa_1@itsyouonline',[user['userGroupId'] for user in cloudspace['acl']])
+            self.assertIn('gig_qa_1@itsyouonline', [user['userGroupId'] for user in cloudspace['acl']])
+
+        self.log('%s ENDED' % self._testID)
+
+
+class CloudspaceActions(OVC_BaseTest):
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: for this class please make sure the tear down for the parent class is not called till all the tests finished

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 9]
```diff
@@ -138,6 +148,259 @@ def test003_create_cloudspace_with_different_limitaions(self, type, factor):
...
+        self.assertEqual(cloudspace['resourceLimits']['CU_M'], CU_M+1)
+        self.assertEqual(cloudspace['resourceLimits']['CU_C'], CU_C-1)
+        self.assertEqual(cloudspace['resourceLimits']['CU_D'], CU_D+1)
+        self.assertEqual(cloudspace['resourceLimits']['CU_I'], CU_I-1)
+        self.assertEqual(cloudspace['resourceLimits']['CU_NP'], CU_NP+1)
+
+        self.log('%s ENDED' % self._testID)
+
+    @unittest.skip("https://github.com/openvcloud/0-templates/issues/74")
+    def test002_update_non_exist_cloudspace(self):
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: integrate this test case with the one above it

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 10]
```diff
@@ -138,6 +148,259 @@ def test003_create_cloudspace_with_different_limitaions(self, type, factor):
...
+        CU_M = cloudspace["maxMemoryCapacity"]
+        temp_actions = {'vdc': {'actions': ['update'], 'args': {"maxMemoryCapacity": CU_M+1}, 'service': fake_cloudspace}}
+        self.cloudspaces = {fake_cloudspace: {'account': self.acc1, 'maxMemoryCapacity': CU_M}}
+        res = self.create_cs(openvcloud=self.openvcloud, vdcusers=self.vdcusers, accounts=self.accounts,
+                             cloudspaces=self.cloudspaces, temp_actions=temp_actions)
+        self.assertEqual(res, "cloudspace doesn't exist")
+
+        self.log('%s ENDED' % self._testID)
+
+    @unittest.skip('https://github.com/openvcloud/0-templates/issues/55')
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: integrate this test case with the above 2

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 11]
```diff
@@ -138,6 +148,259 @@ def test003_create_cloudspace_with_different_limitaions(self, type, factor):
...
+        #. Check that user[U1] added to cloudspace [CS1] sucessfuly , should succeed.
+        #. Delete User [U1] from cloudspace[CS1], should succeed.
+        #. Check that user[U1] deleted from cloudspace [CS1] sucessfuly , should succeed.
+        """
+
+        self.log('%s STARTED' % self._testID)
+
+        self.log("Create user [U1] ,should succeed.")
+        user = self.random_string()
+        vdcuser = {user: {'openvcloud': self.openvcloud,
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: why self.vdcuser is not used?

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 12]
```diff
@@ -138,6 +148,259 @@ def test003_create_cloudspace_with_different_limitaions(self, type, factor):
...
+    def test005_add_and_delete_user_to_cloudspace(self):
+        """ ZRT-OVC-012
+        *Test case for adding and deleting user to cloudspace *
+
+        **Test Scenario:**
+
+        #. Create an cloudspace[CS1], should succeed. 
+        #. Create new user[U1], should succeed.
+        #. Add User[U1] to cloudspace[CS1] , should succeed.
+        #. Check that user[U1] added to cloudspace [CS1] sucessfuly , should succeed.
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: remove this check msg and the one after it

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 13]
```diff
@@ -138,6 +148,259 @@ def test003_create_cloudspace_with_different_limitaions(self, type, factor):
...
+
+        self.log('%s ENDED' % self._testID)
+
+    def test005_add_and_delete_user_to_cloudspace(self):
+        """ ZRT-OVC-012
+        *Test case for adding and deleting user to cloudspace *
+
+        **Test Scenario:**
+
+        #. Create an cloudspace[CS1], should succeed.
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: please add and delete fake users .. using user_add and user_delete actions

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 14]
```diff
@@ -138,6 +148,259 @@ def test003_create_cloudspace_with_different_limitaions(self, type, factor):
...
+        """ ZRT-OVC-013
+        *Test case for adding and deleting portforward on cloudspaces. *
+
+        **Test Scenario:**
+
+        #. Create an cloudspace[CS1], should succeed. 
+        #. Create portforward for created cloudspace, should succeed.
+        #. Check that portforwad  added to cloudspace [CS1] sucessfuly , should succeed.
+        #. Delete the portforward from [CS1], should succeed.
+        #. Check that the portforward  deleted from cloudspace [CS1] sucessfuly , should succeed.
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: --> has been deleted

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 15]
```diff
@@ -138,6 +148,259 @@ def test003_create_cloudspace_with_different_limitaions(self, type, factor):
...
+    @unittest.skip("https://github.com/openvcloud/0-templates/issues/80")
+    def test006_add_and_delete_portforward(self):
+        """ ZRT-OVC-013
+        *Test case for adding and deleting portforward on cloudspaces. *
+
+        **Test Scenario:**
+
+        #. Create an cloudspace[CS1], should succeed. 
+        #. Create portforward for created cloudspace, should succeed.
+        #. Check that portforwad  added to cloudspace [CS1] sucessfuly , should succeed.
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: --> has been added

Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: no need for this check message and the one after

File: tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py [hunk 16]
```diff
@@ -32,37 +31,37 @@ def setUp(self):
...
 
         **Test Scenario:**
-        #. Create an cloudspace with providing non existing parameter, should fail.
-        #. Create an cloudspace with non-existing account , should fail.
-        #. Create an cloudspace without providing account parameter, should fail.
-        #. Create an cloudspace with providing non existing user, should fail
+        #. Create a cloudspace with providing non existing parameter, should fail.
+        #. Create a cloudspace with non-existing account , should fail.
+        #. Create a cloudspace without providing account parameter, should fail.
+        #. Create a cloudspace with providing non existing user, should fail
```
Comment (john-kheir) [inline tests/integration_tests/testsuite/ovc_tests/a_basic/cloudspaces_tests.py]: remove this step

**Comment by Dinaamagdy** [comment]
----------------------------------------------------------------------
Ran 8 tests in 326.283s

OK (SKIP=5)

**Pull request merged**